# QCatch
Quality Control downstream of alevin-fry / simpleaf.

**QCatch** is a Python package designed to streamline quality control for single-cell sequencing data quantified by [alevin-fry](https://github.com/COMBINE-lab/alevin-fry) or [simpleaf](https://github.com/COMBINE-lab/simpleaf). It provides a comprehensive web-based quality control report, enabling researchers to:

- Summarize key **quality metrics** for single-cell sequencing datasets.
- Perform **cell calling** to identify high-quality cells.
- Generate interactive **visualizations** to support downstream analysis and interpretation.

QCatch is built to simplify the quality control process, making it easier for researchers to assess data quality and make informed decisions for further analysis.

## Installation

### Bioconda
You can install using [Conda](http://anaconda.org/)
from [Bioconda](https://bioconda.github.io/).

```bash
conda install -c bioconda qcatch 
```

### PyPI

You can also install `QCatch` from [PyPI](https://pypi.org/project/qcatch/) using `pip`:

```bash
pip install qcatch
```

This is the recommended method if you're not using Conda or want to include `QCatch` in a Python virtual environment.

## Basic Usage
Specify the path to the folder containing quantification results generated by `alevin-fry` or `simpleaf`. QCatch will scan the folder, evaluate the data quality, and generate an HTML report that can be viewed directly in your browser.

```bash
qcatch \
    --input path/to/quants/result/folder \
    --output path/to/QC/output/folder \
    --chemistry 10X_3p_v3 
```
We highly recommend specifying the chemistry used in your experiment, although the tool can run without it. By default, QCatch will assume the settings for 10X 3' v2 and v3 chemistry.

## Command-Line Arguments

| Flag | Short | Type | Description |
|------|-------|------|-------------|
| `--input`  | `-i` | `str` (Required) | Path to the input directory containing the quant output files. |
| `--output` | `-o` | `str`(Optional)  | Path to the output directory. **Default**: Same directory as input. |
| `--chemistry` | `-c` | `str`(Optional but recommend) | Specifies the chemistry used in the experiment, determining the range for the `empty_drops` step. **Options**: `'10X_3p_v2'`, `'10X_3p_v3'`, `'10X_3p_v4'`, `'10X_3p_LT'`,`'10X_5p_v3'`,`'10X_HT'`. **Default**: Will use the range for `'10X_3p_v2'` and `'10X_3p_v3'`. |
| `--n_partitions` | `-n` | `int` (Optional) | Number of partitions (max number of barcodes to consider for ambient estimation). Skip this step if you already specified `--chemistry`. Only use `--n_partitions` when your experiment uses a custom chemistry not listed in the predefined chemistry options.|
| `--gene_id2name_dir` | `-g` | `str` (Optional) | Directory containing the `gene_id2name` file for converting Ensembl gene IDs to gene names. The file must be a CSV with two columns: `gene_id` and `gene_name`. If not provided, mitochondria plots requiring gene names will not be displayed. **Default**: `None`. |
| `--verbose` | `-v` | `flag` (Optional) | Enable verbose logging with debug-level messages. |
| `--overwrite_h5ad` | `-o` | `flag` (Optional) |If enabled, `QCatch` will overwrite the original `.h5ad` file in place by appending cell filtering results to `anndata.obs`. No existing data or cells will be removed; only additional metadata columns are added.|

